spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -67,46 +67,48 @@ class DeltaHistoryManager(
    * chronological order.
    */
   def getHistory(limitOpt: Option[Int]): Seq[DeltaHistory] = {
+    val snapshot = deltaLog.update()
     val listStart = limitOpt.map { limit =>
-      math.max(deltaLog.update().version - limit + 1, 0)
+      math.max(snapshot.version - limit + 1, 0)
     }.getOrElse(getEarliestDeltaFile(deltaLog))
-    getHistory(listStart)
+    getHistory(listStart, end = Some(snapshot.version))
   }
 
   /**
-   * Get the commit information of the Delta table from commit `[start, end]`. If `end` is `None`,
-   * we return all commits from start to now.
+   * Get the commit information of the Delta table from commit `[start, end)`.
+   *
+   * @param isRangeICT Whether the commit range is ICT-enabled.
+   *                   If `true`, all commits in the range must have ICTs and
+   *                   the timestamp returned for each commit will be the ICT.
+   *                   If `false`, the file modification time will be used as the timestamp.
    */
-  def getHistory(
+  private[delta] def getHistoryImpl(
       start: Long,
-      endOpt: Option[Long] = None): Seq[DeltaHistory] = {
+      end: Long,
+      isRangeICT: Boolean,
+      snapshot: Snapshot): Seq[DeltaHistory] = {
     import org.apache.spark.sql.delta.implicits._
     val conf = getSerializableHadoopConf
     val logPath = deltaLog.logPath.toString
-    val currentSnapshot = deltaLog.unsafeVolatileSnapshot
-    val (snapshotForCommitFileProvider, end) = endOpt match {
-      case Some(end) if currentSnapshot.version >= end =>
-        // Use the cache snapshot if it's fresh enough for the [start, end] query.
-        (currentSnapshot, end)
-      case _ =>
-        // Either end doesn't exist or the currently cached snapshot isn't new enough to satisfy it.
-        val newSnapshot = deltaLog.update()
-        val endVersion = endOpt.getOrElse(newSnapshot.version).min(newSnapshot.version)
-        (newSnapshot, endVersion)
-    }
-    val commitFileProvider = DeltaCommitFileProvider(snapshotForCommitFileProvider)
+    val commitFileProvider = DeltaCommitFileProvider(snapshot)
     // We assume that commits are contiguous, therefore we try to load all of them in order
-    val info = spark.range(start, end + 1)
+    val info = spark.range(start, end)
       .mapPartitions { versions =>
         val logStore = LogStore(SparkEnv.get.conf, conf.value)
         val basePath = new Path(logPath)
         val fs = basePath.getFileSystem(conf.value)
         versions.flatMap { commit =>
           try {
             val deltaFile = commitFileProvider.deltaFile(commit)
-            val ci = DeltaHistoryManager.getCommitInfo(logStore, deltaFile, conf.value)
-            val metadata = fs.getFileStatus(deltaFile)
-            Some(ci.withTimestamp(metadata.getModificationTime))
+            val ci = DeltaHistoryManager
+              .getCommitInfoOpt(logStore, deltaFile, conf.value)
+              .getOrElse(CommitInfo.empty(Some(commit)))
```

Comment:
Using an empty CommitInfo here will result in throwing [missingCommitTimestamp](https://github.com/delta-io/delta/blob/01c0ef91565d9a3e06fb9f62295f7c80fba09351/spark/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L1203) error here instead of `missingCommitInfo` error. I'm wondering if the latter would be better instead 🤔

Reply:
`missingCommitInfo` is definitely the better error message in that case. Restructured the code to fix this.